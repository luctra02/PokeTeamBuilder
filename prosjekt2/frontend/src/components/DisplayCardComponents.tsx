import CardComponent from '../components/CardComponent';
import { KeyboardEvent, useState,useEffect } from 'react';
import { getTeamSize } from '../utils/teamFunctions';
import { Pagination } from '@mui/material';
import fetchPokemonList from '../assets/PokemonDatabase';
import { useLocation, useNavigate } from 'react-router-dom';

interface PokemonObject {
  id: number;
  name: string;
  image: string;
  types: string[];
  weight: number;
  height: number;
  baseStats: {
    attack: number;
    defense: number;
    hp: number;
    speed: number;
    specialattack: number;
    specialdefense: number;
  };
}


function DisplayCardComponents() {
  if(!localStorage.getItem("PokemonDatabase")){
    fetchPokemonList();
  }
  const pokemonDatabase = localStorage.getItem("PokemonDatabase")
  const pokemonArray: PokemonObject[] = pokemonDatabase ? JSON.parse(pokemonDatabase) : [];
  const [, setCount] = useState(getTeamSize());

  function updateCount(count: number) {
    setCount(count);
  }

  const itemsPerPage = 16;

  const [pageNumber, setPageNumber] = useState(1);
  const [searchedArray, setSearchedArray] = useState(pokemonArray);
  const [totalSearchedItems, setTotalSearchedItems] = useState(0);
  const location = useLocation();
  const navigate = useNavigate();
  const totalPages = Math.ceil(totalSearchedItems / itemsPerPage);

  function changePage(newPage: number) {
    if (newPage >= 1 && newPage <= totalPages) {
      setPageNumber(newPage);
    }
  }

  useEffect(() => {
    if (location.state?.searchTerm != null) {
      const searchTerm = location.state.searchTerm.toLowerCase();
      const filteredArray = pokemonArray.filter((pokemon) => pokemon.name.includes(searchTerm));
      setSearchedArray(filteredArray);
      setTotalSearchedItems(filteredArray.length);
      setPageNumber(1);
    } else {
      setSearchedArray(pokemonArray);
      setTotalSearchedItems(pokemonArray.length);
    }
  }, [location.state?.searchTerm]);


  function changeToDetailPage(pokemon: PokemonObject) {
    navigate(`/pokemonInfo/${pokemon.id}`, { state: { pokemon } });
  }

  const handleEnterPress = (event : KeyboardEvent, pokemon: PokemonObject) => {
    // Check if the key pressed is 'Enter'
    if (event.key === 'Enter') {
      changeToDetailPage(pokemon);
    }
};
  const startItem = (pageNumber - 1) * itemsPerPage;
  const endItem = Math.min(startItem + itemsPerPage, totalSearchedItems);

  return (
    <>
      <div className="pokemonDisplayBox">
        {searchedArray.slice(startItem, endItem).map((pokemon) => (
          <div className="pokemonDisplayButton" tabIndex={0} key={pokemon.id} onClick={() => changeToDetailPage(pokemon)} onKeyDown={(event) => handleEnterPress(event, pokemon)}>
            <CardComponent
              pokemonObject={{
                id: pokemon.id,
                name: pokemon.name,
                image: pokemon.image,
                types: pokemon.types,
                weight: pokemon.weight,
                height: pokemon.height,
                baseStats: pokemon.baseStats
              }}
              updateCount = {updateCount}
            />
          </div>
        ))}
      </div>
      <div className="pageSelector">
        <Pagination count={totalPages} page={pageNumber} onChange={(_event, value) => changePage(value)} />
      </div>
    </>
  );
}

export default DisplayCardComponents;
