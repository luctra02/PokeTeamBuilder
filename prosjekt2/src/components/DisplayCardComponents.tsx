import CardComponent from "../components/CardComponent"
import { useState } from "react"
import pokemonArray  from "../assets/PokemonList"
import { useNavigate } from "react-router-dom";
import { getTeamSize } from "../utils/teamFunctions";

interface PokemonObject {
  num: number;
  sprite: string;
  types: string[];
  key: string;
  weight: number;
  height: number;
  baseStats: number[]
}

function DisplayCardComponents() {
  const [_count, setCount] = useState(getTeamSize()); 

  function updateCount(count: number) {
    setCount(count);
  }

  const itemsPerPage = 15;
  const totalItems = pokemonArray.length;
  const totalPages = Math.ceil(totalItems / itemsPerPage);

  const [pageNumber, setPageNumber] = useState(1);

  function changePage(newPage: number) {
    if (newPage >= 1 && newPage <= totalPages) {
      setPageNumber(newPage);
    }
  }

  function generatePageButtons() {
    const pageButtons = [];
    const numToShow = 2;

    for (let i = pageNumber - numToShow; i <= pageNumber + numToShow; i++) {
      if (i >= 1 && i <= totalPages) {
        pageButtons.push(
          <button key={i} onClick={() => changePage(i)}>
            {i}
          </button>
        );
      }
    }

    return pageButtons;
  }

  const startItem = (pageNumber - 1) * itemsPerPage;
  const endItem = Math.min(startItem + itemsPerPage, totalItems);

  //navigate to detailspage for a selected pokemon
  const navigate = useNavigate(); 

  function changeToDetailPage(pokemon: PokemonObject) {
    navigate(`/pokemonInfo/${pokemon.num}`, { state: { pokemon } });
  }

  return (
    
    <>
      <div className="pokemonDisplayBox">
        {pokemonArray.slice(startItem, endItem).map((pokemon) => (
          <div className="pokemonDisplayButton" key={pokemon.num} onClick={() => (changeToDetailPage(pokemon))}>
          <CardComponent
            pokemonObject={{id:pokemon.num, name:pokemon.key, image:pokemon.sprite, types:pokemon.types}}
            updateCount={updateCount}
          /></div>
        ))}
      </div>
      <div className="pageSelector">
        <button onClick={() => changePage(pageNumber - 1)}>Previous page</button>
          {generatePageButtons()}
          <button onClick={() => changePage(pageNumber + 1)}>Next page</button>
      </div>
    </>
  );
}

export default DisplayCardComponents;