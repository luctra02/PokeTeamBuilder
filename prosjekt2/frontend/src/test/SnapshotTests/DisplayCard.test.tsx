import { render, screen, waitFor } from '@testing-library/react';
import { allMocks as mocks} from '../mocks/mockQueries';
import { testPageRender } from '../testUtils';
import DisplayCardComponents from '../../components/DisplayCardComponents';



describe('renders pokemon', () => {
    test('matches the snapshot', async () => {
        const { asFragment } = testPageRender(<DisplayCardComponents />, {mocks})
  
        await waitFor(() => {
            expect(screen.getByText(/Pikachu/i)).toBeInTheDocument();
            expect(screen.getByText(/Staraptor/i)).toBeInTheDocument();

        });
  
        expect(asFragment()).toMatchSnapshot();
    });
  }); 
