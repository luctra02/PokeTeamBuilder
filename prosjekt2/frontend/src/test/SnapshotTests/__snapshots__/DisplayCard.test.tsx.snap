// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders pokemon > matches the snapshot 1`] = `
<DocumentFragment>
  <section>
    <section
      class="pokemonDisplayBox"
    >
      <article
        class="pokemonDisplayButton"
        tabindex="0"
      >
        <article
          class="pokemonCard"
        >
          <h2>
            Pikachu
          </h2>
          <img
            alt=""
            src="https://play.pokemonshowdown.com/sprites/ani/pikachu.gif"
          />
          <div
            class="pokemonID"
          >
            #0025
          </div>
          <div
            class="pokeType"
          >
            <section
              class="pokeType PokeTypeBox"
              style="background-color: rgb(248, 208, 48);"
            >
              Electric
            </section>
          </div>
          <div>
            Loading...
          </div>
        </article>
      </article>
      <article
        class="pokemonDisplayButton"
        tabindex="0"
      >
        <article
          class="pokemonCard"
        >
          <h2>
            Staraptor
          </h2>
          <img
            alt=""
            src="https://play.pokemonshowdown.com/sprites/ani/staraptor.gif"
          />
          <div
            class="pokemonID"
          >
            #0398
          </div>
          <div
            class="pokeType"
          >
            <section
              class="pokeType PokeTypeBox"
              style="background-color: rgb(168, 168, 120);"
            >
              Normal
            </section>
            <section
              class="pokeType PokeTypeBox"
              style="background-color: rgb(168, 144, 240);"
            >
              Flying
            </section>
          </div>
          <div>
            Loading...
          </div>
        </article>
      </article>
    </section>
    <div
      class="pageSelector"
    >
      <nav
        aria-label="pagination navigation"
        class="MuiPagination-root MuiPagination-text css-1oj2twp-MuiPagination-root"
      >
        <ul
          class="MuiPagination-ul css-wjh20t-MuiPagination-ul"
        >
          <li>
            <button
              aria-label="Go to previous page"
              class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-disabled MuiPaginationItem-previousNext css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                data-testid="NavigateBeforeIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                />
              </svg>
            </button>
          </li>
          <li>
            <button
              aria-current="true"
              aria-label="page 1"
              class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-selected MuiPaginationItem-page css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
              tabindex="0"
              type="button"
            >
              1
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </li>
          <li>
            <button
              aria-label="Go to next page"
              class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-disabled MuiPaginationItem-previousNext css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                data-testid="NavigateNextIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                />
              </svg>
            </button>
          </li>
        </ul>
      </nav>
    </div>
  </section>
</DocumentFragment>
`;
